### ISSUE TITLE
Odd MLX Adam Optimizer Behavior in Simple Benchmark Example

### ISSUE THREAD

**Issue by sck-at-ucy** [ml-explore/mlx#504]
Here are two codes, one in PyTorch and one in MLX which use the Adam Optimizer to fit a spline to a sine wave. I like this simple benchmark because you can visualize the behavior.  The Torch optimization works well and converges monotonically without being too sensitive on the learning rate. I could not get the MLX Adam optimizer to behave well. With this definition of the loss 

```
def loss_fun(model, x, y_true):
    y_pred = model.forward(x)
    return mx.mean((y_pred - y_true) ** 2)
```

after a few epochs the code starts producing `nan`. If I switch to 

```
def loss_fun(model, x, y_true):
    y_pred = model.forward(x)
    return mx.max((y_pred - y_true) ** 2)
```

the code avoids  `nan` but the control points are moved to the true positions in an erratic non-monotonic way. 

Here are the two codes for comparison. It is worth allowing the frames to be saved to see what is going on, even though that slows down execution of course. 

### TORCH CODE 
```
import torch
import torch.optim as optim
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import os

# TEST CASE OF OPTIMIZERS
# USE TORCH TO A FIT A SIMPLE SPLINE TO A SINE WAVE 

# Simple cubic spline interpolation function (1D)
def cubic_spline(x, control_points):
    # This is a simplified version for illustration purposes.
    # In a real scenario, you would implement a more robust spline calculation.
    n = len(control_points) - 1
    intervals = 1 / n
    index = (x * n).long()
    index = torch.clamp(index, 0, n - 1)
    t = x * n - index.float()
    a0 = control_points[index]
    a1 = control_points[index + 1]
    t2 = t * t
    t3 = t2 * t
    return (2 * t3 - 3 * t2 + 1) * a0 + (t3 - 2 * t2 + t) * (a1 - a0) + (-2 * t3 + 3 * t2) * a1


# Target function to fit (example: sine wave)
def target_function(x):
    return torch.sin(2 * np.pi * x)


# Control points initialization
num_control_points = 100
control_points = torch.linspace(0, 1, num_control_points, requires_grad=True)
control_points_PRE = torch.linspace(0, 1, num_control_points)


x = torch.linspace(0, 1, 100)
y_pred = cubic_spline(x, control_points_PRE)
y_true = target_function(x)

# Plot the results
plt.plot(x, y_true, label='Target Function')
plt.scatter(np.linspace(0, 1, num_control_points), control_points_PRE, color='red', label='Control Points')
plt.plot(x, y_pred, label='Spline Fit')
plt.legend()
plt.show()
plt.close()

# Optimizer
optimizer = optim.Adam([control_points], lr=0.01)


movie_dir = './MOVIEW'
os.makedirs(movie_dir, exist_ok=True)
# Training loop
num_epochs = 500

for epoch in range(num_epochs):
    optimizer.zero_grad()

    # Generate points for interpolation
    x = torch.linspace(0, 1, 100)
    y_pred = cubic_spline(x, control_points)
    y_true = target_function(x)

    # Compute loss (mean squared error)
    loss = torch.mean((y_pred - y_true) ** 2)

    # Backward pass and optimization
    loss.backward()
    optimizer.step()

    # Plot the results
    plt.plot(x.detach().numpy(), y_true.detach().numpy(), label='Target Function')
    plt.scatter(np.linspace(0, 1, num_control_points), control_points.detach().numpy(), color='red',
                label='Control Points')
    plt.plot(x.detach().numpy(), y_pred.detach().numpy(), label='Spline Fit')
    plt.legend()

    # Save each plot with a unique filename
    plt.savefig(f'{movie_dir}/spline_fit_epoch_{epoch:05d}.png', format='png', dpi=300)
    plt.close()  # Close the plot to prevent overlap

    # Print loss every 100 epochs
    if epoch % 100 == 0:
        print(f'Epoch {epoch}, Loss: {loss.item()}')

# Plot the results
plt.plot(x.detach().numpy(), y_true.detach().numpy(), label='Target Function')
#plt.scatter(np.linspace(0, 1, num_control_points), control_points.detach().numpy(), color='red', label='Control Points')
plt.plot(x.detach().numpy(), y_pred.detach().numpy(), label='Spline Fit')
plt.legend()
plt.show()

```

### MLX CODE

```
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim
import numpy as np
import matplotlib.pyplot as plt
import os

# TEST CASE OF OPTIMIZERS
# USE MLX TO A FIT A SIMPLE SPLINE TO A SINE WAVE

# Define the model class
class SplineModel(nn.Module):
    def __init__(self, num_control_points):
        super(SplineModel, self).__init__()
        # Initialize control points as trainable parameters
        self.control_points = mx.array(mx.linspace(0, 1, num_control_points))

    def forward(self, x):
        return cubic_spline(x, self.control_points)

# Define the loss function
def loss_fun(model, x, y_true):
    y_pred = model.forward(x)
    return mx.mean((y_pred - y_true) ** 2)

# Simple cubic spline interpolation function (1D)
# Define the simple cubic spline interpolation function (1D)
def cubic_spline(x, control_points):
    n = len(control_points) - 1
    intervals = 1 / n
    index = (x * n).astype(mx.int32)
    index = mx.clip(index, 0, n - 1)
    t = x * n - index
    a0 = control_points[index]
    a1 = control_points[index + 1]
    t2 = t * t
    t3 = t2 * t
    return (2 * t3 - 3 * t2 + 1) * a0 + (t3 - 2 * t2 + t) * (a1 - a0) + (-2 * t3 + 3 * t2) * a1


# Target function to fit (example: sine wave)
def target_function(x):
    return mx.sin(2 * mx.pi * x)

# Initialize the model
num_control_points = 100
model = SplineModel(num_control_points)

# Optimizer setup
optimizer = optim.Adam(learning_rate=0.01)

# Create a directory for the movie frames
movie_dir = './MOVIEW'
os.makedirs(movie_dir, exist_ok=True)

# Training loop
num_epochs = 500
#clip_value = 1.0  # Maximum allowed gradient value
for epoch in range(num_epochs):
    # Generate points for interpolation
    x = mx.linspace(0, 1, num_control_points)
    y_true = target_function(x)

    loss_fn = lambda: loss_fun(model,x,y_true)
    loss, grads = nn.value_and_grad(model, loss_fn)()

    optimizer.update(model, grads)

    # Compute the new parameters
    mx.eval(model.parameters(), optimizer.state)

    # Plot the results
    x_np, y_true_np = np.array(x), np.array(y_true)
    y_pred_np = np.array(model.forward(x))
    control_points_np = np.array(model.control_points)

    plt.plot(x_np, y_true_np, label='Target Function')
    plt.scatter(np.linspace(0, 1, num_control_points), control_points_np, color='red', label='Control Points')
    plt.plot(x_np, y_pred_np, label='Spline Fit')
    plt.legend()
    plt.savefig(f'{movie_dir}/spline_fit_epoch_{epoch:05d}.png', format='png', dpi=300)
    plt.close()

    if epoch % 100 == 0:
        print(f'Epoch {epoch}, Loss: {loss.item()}')

# Plot the final results
plt.plot(x_np, y_true_np, label='Target Function')
plt.scatter(np.linspace(0, 1, num_control_points), control_points_np, color='red', label='Control Points')
plt.plot(x_np, y_pred_np, label='Spline Fit')
plt.legend()
plt.show()
```

**Comment by awni** [comment]
There is a bug here. I simplified your code to a few lines which shows the problem:

```python
import mlx.core as mx

def loss_fun(y_pred, y_true):
    return mx.sum((y_pred - y_true) ** 2)
      
y_pred = mx.array([0], dtype=mx.float32)
y_true = mx.array([0], dtype=mx.float32)
loss, grads = mx.value_and_grad(loss_fun)(y_pred, y_true)
```

Outputs
```
array([nan], dtype=float32)
```

**Comment by awni** [comment]
I think it was introduced in #475 when we switched the Power VJP...

**Comment by awni** [comment]
For now a simple fix is to replace `**2` with `.square()`. I will push a fix soon.

**Comment by awni** [comment]
FIx in #505, your script converges nicely for me now. Thanks for catching that and sorry for the bug!

**Comment by sck-at-ucy** [comment]
Great thank you. You guys are quick 😊
